typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+    return "valid";
+  } else if (typeof arg === "function") {
+    return undefined;
+  }
+}
+```
+
+Going deeper using Unions it is possible to construct data type analog.
+It is called discriminated Unions.
+The idea is to store special property for each variant which will allow switching on it.
+
+```typescript
+type Result = 
+  | {type: "error"; message: string}
+  | {type: "result"; value: number}
+
+const resultInterpreter = (result: Result): string => {
+  switch (result.type) {
+    case "error":
+      return result.message;
+    case "result":
+      return `valid ${result.value.toString()}`;
+    default:
+      return "impossible";
+  }
+};
+```
+
+This is the very simple examples of working with ADT like structures in TypeScript.
+Using type parameters you are also able to make some more complex things like [such](https://github.com/pfgray/ts-adt) library for ADTs.
+But one of the most useful features of Haskell connected with ADTs is pattern matching.
+In allows you to disclose you structure comparing it with some pattern and based on this condition execute some code.
+There is [no such functionality](https://github.com/tc39/proposal-pattern-matching) in TypeScript but [ts-pattern](https://github.com/gvergnaud/ts-pattern) will bring it to you.
+Example from repository:
+
+```typescript
+type Data =
+  | { type: 'text'; content: string }
+  | { type: 'img'; src: string };
+
+type Result =
+  | { type: 'ok'; data: Data }
+  | { type: 'error'; error: Error };
+
+const result: Result = ...;
+
+return match(result)
+  .with({ type: 'error' }, (res) => `<p>Oups! An error occured</p>`)
+  .with({ type: 'ok', data: { type: 'text' } }, (res) => `<p>${res.data.content}</p>`)
+  .with({ type: 'ok', data: { type: 'img', src: select() } }, (src) => `<img src=${src} />`)
+  .exhaustive();
+```
+
+### Immutability
+
+TypeScript has two base primitives to work with immutability: const and readonly.
+The first one is used to prevent variable reference change.
+And the second one for properties immutability.
+
+```typescript
+const a = 1;
+a = 2; // error
+
+interface A {
+  readonly x: number;
+}
+const a: A = { x: 1 };
+a.x = 12; // error
+```
+
+Nevertheless, you can easily see problems here.
+Const and readonly only block the reference change, but make nothing about values.
+With `const a = [1, 2, 3]` or `readonly x: number[]` you still will be able to change the content of array.
+Let's introduce more cases and how you can deal with them in TypeScript.
+
+```typescript
+type A = {
+  x: number;
+  y: number;
+};
+
+type ImmutableA = Readonly<A>; 
+const a: ImmutableA = { x: 1, y: 2 };
+a.x = 2; // error
+a.y = 1; // error
+
+
+const arr: ReadonlyArray<number> = [1, 2, 3];
+arr[0] = 4; // error
+
+
+interface A {
+    readonly [x: string]: number;
+};
+
+const a: A = { "x": 1, "y": 2 };
+a["x"] = 2; // error
+a["y"] = 1; // error
+```
+
+Also be careful passing object with readonly fields to function as it may cause unexpected behaviour.
```

Comment:
As I understand that is because of scope and that is a problem. So, function know nothing about `arg` type beyond itself. It know only that `arg: { x: number }`. But maybe not, will try to understand it more